compiler/optimizations/remoteValueForwarding.cpp

```diff
@@ -579,3 +596,94 @@ static bool isSafeToDeref(Map<Symbol*, Vec<SymExpr*>*>& defMap,
   return retval;
 }
 
+static bool computeDotLocale(Symbol* sym) {
+  INT_ASSERT(sym->isRef());
+
+  DotInfo* info = dotLocaleMap[sym];
+
+  if (info == NULL) {
+    return false; // No uses for this symbol
+  } else if (info->finalized) {
+    return info->usesDotLocale;
+  }
+
+  bool retval = false;
+
+  while (!info->todo.empty() && !retval) {
+    SymExpr* use = info->todo.back();
+    info->todo.pop_back();
+
+    CallExpr* call = toCallExpr(use->parentExpr);
+    INT_ASSERT(call);
+    CallExpr* parent = toCallExpr(call->parentExpr);
+
+    if (call->isResolved()) {
+      ArgSymbol* form = actual_to_formal(use);
+      if (form->isRef() && computeDotLocale(form)) retval = true;
+    } else if (parent && isMoveOrAssign(parent)) {
+      // If something copies this reference, we need to see if the new
+      // reference also uses .locale
+      Symbol* LHS = toSymExpr(parent->get(1))->var;
+      if (call->isPrimitive(PRIM_ADDR_OF) ||
+          call->isPrimitive(PRIM_SET_REFERENCE)) {
+        if (computeDotLocale(LHS)) retval = true;
+      } else if (call->isPrimitive(PRIM_GET_MEMBER) ||
+                 call->isPrimitive(PRIM_GET_MEMBER_VALUE) ||
+                 call->isPrimitive(PRIM_GET_SVEC_MEMBER) ||
+                 call->isPrimitive(PRIM_GET_SVEC_MEMBER_VALUE)) {
+        // Check if the field is used in a .locale
+        if (LHS->isRef() && use == call->get(2) && computeDotLocale(LHS)) {
+          retval = true;
+        }
+      } else if (call->isPrimitive(PRIM_WIDE_GET_LOCALE) ||
+                 call->isPrimitive(PRIM_WIDE_GET_NODE)) {
+        retval = true;
+      }
+    } else if (call->isPrimitive(PRIM_MOVE)) {
+      SymExpr* LHS = toSymExpr(call->get(1));
+      if (LHS->isRef() && use != LHS && computeDotLocale(LHS->var)) retval = true;
+    } else if (call->isPrimitive(PRIM_SET_MEMBER) && use == call->get(3)) {
+      // See if the field we're setting may use .locale
+      // This could be improved by only looking at the base object's instance,
+      // but that's much more complicated.
+      SymExpr* member = toSymExpr(call->get(2));
+      if (member->isRef() && computeDotLocale(member->var)) {
+        retval = true;
+      }
+    } else if (call->isPrimitive(PRIM_SET_SVEC_MEMBER)) {
+      // Not sure how to handle the homogeneous tuple case yet, so to be safe
+      // simply return true;
```

Comment:
I could do that, but I'm not sure if it's worth it for this PR. I can at least update the comment expressing that view.